src/Tokenizer/Analyzer/ClassesAnalyzer.php

```diff
@@ -0,0 +1,120 @@
+<?php
+
+/*
+ * This file is part of PHP CS Fixer.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *     Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PhpCsFixer\Tokenizer\Analyzer;
+
+use PhpCsFixer\Tokenizer\Analyzer\Analysis\TypeAnalysis;
+use PhpCsFixer\Tokenizer\Tokens;
+use PhpCsFixer\Tokenizer\TokensAnalyzer;
+
+/**
+ * @internal
+ */
+final class ClassesAnalyzer
+{
+    /**
+     * @param int $classIndex
+     *
+     * @return array Contains information about the class being analyzed
+     */
+    public function getClassDefinition(Tokens $tokens, $classIndex)
+    {
+        $openIndex = $tokens->getNextTokenOfKind($classIndex, ['{']);
+        $prevIndex = $tokens->getPrevMeaningfulToken($classIndex);
+        $startIndex = $tokens[$prevIndex]->isGivenKind([T_FINAL, T_ABSTRACT]) ? $prevIndex : $classIndex;
+
+        $extends = false;
+        $implements = false;
+        $anonymousClass = false;
+
+        if (!$tokens[$classIndex]->isGivenKind(T_TRAIT)) {
+            $extends = $tokens->findGivenKind(T_EXTENDS, $classIndex, $openIndex);
+            $extends = 0 !== \count($extends) ? $this->getClassInheritanceInfo($tokens, key($extends), 'numberOfExtends') : false;
+
+            if (!$tokens[$classIndex]->isGivenKind(T_INTERFACE)) {
+                $implements = $tokens->findGivenKind(T_IMPLEMENTS, $classIndex, $openIndex);
+                $implements = \count($implements) ? $this->getClassInheritanceInfo($tokens, key($implements), 'numberOfImplements') : false;
+                $tokensAnalyzer = new TokensAnalyzer($tokens);
+                $anonymousClass = $tokensAnalyzer->isAnonymousClass($classIndex);
+            }
+        }
+
+        return [
```

Comment:
What about adding `<Name>Analysis` classes instead of associative arrays here and below in `getClassInheritanceInfo`? Would that be the next step (separate PR)?

Reply:
I don't get it, can you please explain a bit more ?

Reply:
In some places (like [here](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/v2.16.3/src/Tokenizer/Analyzer/NamespaceUsesAnalyzer.php#L39)) instead of returning array we return an object (like [this](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/v2.16.3/src/Tokenizer/Analyzer/Analysis/NamespaceUseAnalysis.php)).

The returned objects are very simple and make using it easier than using an array.

I believe in the new code we want to use object approach over an array one.

Reply:
Even more, the existing analysers should start returning objects as well.
- they are marked as internal anyway
- we can provide arrayaccess for compatibility reasons, if really needed

Reply:
Should we do that now or in a future PR ?

Reply:
Ok I did it within this PR in commit https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4362/commits/889f98435d7d3f2251296220b936e4d0a1e96bbe. Let me know if it's ok now.